Quotient/connectionencryptiondata_p.cpp

```diff
@@ -371,26 +376,183 @@ void ConnectionEncryptionData::handleEncryptedToDeviceEvent(
+                    "UPDATE master_keys SET verified=0;"_ls);
+                database.execute(query);
+                database.commit();
+            }
+        }
+
+        auto masterKey = masterKeyQuery.value("key"_ls).toString();
+        auto signature = key.signatures[userId]["ed25519:"_ls % masterKey].toString();
+        if (!ed25519VerifySignature(masterKey, toJson(key), signature)) {
+            qWarning() << "User signing key: failed signature verification" << userId;
+            continue;
+        }
+        auto query = database.prepareQuery("DELETE FROM user_signing_keys WHERE userId=:userId;"_ls);
+        query.bindValue(":userId"_ls, userId);
+        database.execute(query);
+        query = database.prepareQuery("INSERT INTO user_signing_keys(userId, key) VALUES(:userId, :key);"_ls);
+        query.bindValue(":userId"_ls, userId);
+        query.bindValue(":key"_ls, key.keys.values()[0]);
+        database.execute(query);
+    }
+    database.commit();
+    if (q->isUserVerified(q->userId())) {
+        auto query = database.prepareQuery("SELECT key FROM user_signing_keys WHERE userId=:userId;"_ls);
+        query.bindValue(":userId"_ls, q->userId());
+        database.execute(query);
+        query.next();
+        auto userSigningKey = query.value("key"_ls).toString();
+        for (const auto& masterKey : masterKeys) {
+            auto signature = masterKey.signatures[q->userId()]["ed25519:"_ls % userSigningKey].toString();
+            if (!signature.isEmpty()) {
+                if (ed25519VerifySignature(userSigningKey, toJson(masterKey), signature)) {
+                    database.setMasterKeyVerified(masterKey.keys.values()[0]);
+                    emit q->userVerified(masterKey.userId);
+                } else {
+                    qCWarning(E2EE) << "Master key signature verification failed";
+                }
+            }
+        }
+    }
+    for(const auto &[user, keys] : asKeyValueRange(deviceKeys)) {
+        QHash<QString, Quotient::DeviceKeys> oldDevices = this->deviceKeys[user];
+        auto query = database.prepareQuery("SELECT * FROM self_signing_keys WHERE userId=:userId;"_ls);
+        query.bindValue(":userId"_ls, user);
+        database.execute(query);
+        auto selfSigningKey = query.next() ? query.value("key"_ls).toString() : QString();
+        this->deviceKeys[user].clear();
+        selfVerifiedDevices[user].clear();
+        for(const auto &device : keys) {
+            if(device.userId != user) {
                 qWarning(E2EE)
                     << "mxId mismatch during device key verification:"
                     << device.userId << user;
                 continue;
             }
             if (!std::all_of(device.algorithms.cbegin(),
-                             device.algorithms.cend(), isSupportedAlgorithm)) {
+                            device.algorithms.cend(),
+                            isSupportedAlgorithm)) {
```

Comment:
It seems that your editor disregards `.clang-format`.
```suggestion
                             device.algorithms.cend(), isSupportedAlgorithm)) {
```